Allow a request header to be used to provide the auth token

msessa opened the pull request:
This PR is experimental and is meant to address #39 

Introduce the `VAULT_SUPPLIED_TOKEN_HEADER` server environment variable.

The above variable will instruct the server to lookup in the specified request header for the authentication token.
Combined with the already supported `VAULT_URL_DEFAULT` and `VAULT_AUTH_DEFAULT` env vars, it can be used to enable login with no user interaction.

Commit:
Allow a request header to be used to provide the auth token

Commit:
Reindent

djenriquez commented on app/components/Login/Login.jsx:
```diff
@@ -47,6 +47,13 @@ export default class Login extends React.Component {
+            // If a token was supplied in the window.suppliedAuthToken, simulate a login
+            if ( window.suppliedAuthToken ) {
+                this.setState({loginMethodType: 'TOKEN', authToken: window.suppliedAuthToken}, function () {
+                    this.validateToken({keyCode: 13});
```
Nice work. 

In this case, if the token supplied by the header, which we assume as the SSO entity, fails, it would generate the generic "invalid" login. Can we be more explicit that if this step fails, it fails because of the automated login? This will help users understand to look downstream for the resolution.

Commit:
Improve login error message and behaviour

When using server supplied auth token and authentication fails display
a more descriptive error message and failback to the default auth method

Commit:
Reset default auth method in docker-compose

alexunwin commented on app/components/Login/Login.jsx:
```diff
@@ -47,6 +47,13 @@ export default class Login extends React.Component {
+            // If a token was supplied in the window.suppliedAuthToken variable, then simulate a login
+            if ( window.suppliedAuthToken ) {
+                this.setState({loginMethodType: 'TOKEN', authToken: window.suppliedAuthToken}, function () {
+                    this.validateToken({keyCode: 13, tokenSupplied: true});
```
this logic looks like it belongs in the constructor, can we move it there instead.

alexunwin commented on server.js:
```diff
@@ -10,6 +10,7 @@ var routeHandler = require('./src/routeHandler');
 var PORT = 8000;
 var VAULT_URL_DEFAULT = process.env.VAULT_URL_DEFAULT || "";
 var VAULT_AUTH_DEFAULT = process.env.VAULT_AUTH_DEFAULT || "GITHUB";
+var VAULT_SUPPLIED_TOKEN_HEADER = process.env.VAULT_SUPPLIED_TOKEN_HEADER || undefined;
```
No need for the `|| undefined` here since it will be set to `undefined` anyways if it DNE.

alexunwin commented on server.js:
```diff
@@ -95,5 +96,9 @@ app.post('/unwrap', function(req, res) {
+    res.render(path.join(__dirname, '/index.html'),{
+        defaultUrl: VAULT_URL_DEFAULT,
+        defaultAuth: VAULT_AUTH_DEFAULT,
+        suppliedAuthToken: VAULT_SUPPLIED_TOKEN_HEADER ? req.header(VAULT_SUPPLIED_TOKEN_HEADER) : ""
```
No need to set it to a falsy value, req.header() returns undefined if not found which is falsy anyways.

msessa replied:
This is actually needed, otherwise in situations where the env var is not set it will result in a `req.header("")` call which returns the following exception:
```
TypeError: name argument is required to req.get
   at IncomingMessage.header (/app/node_modules/express/lib/request.js:61:11)
   at /app/server.js:102:32
   at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
   at next (/app/node_modules/express/lib/router/route.js:131:13)
   at Route.dispatch (/app/node_modules/express/lib/router/route.js:112:3)
   at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
   at /app/node_modules/express/lib/router/index.js:277:22
   at param (/app/node_modules/express/lib/router/index.js:349:14)
   at param (/app/node_modules/express/lib/router/index.js:365:14)
   at Function.process_params (/app/node_modules/express/lib/router/index.js:410:3)
```

alexunwin commented on app/components/Login/Login.jsx:
```diff
@@ -113,7 +120,10 @@ export default class Login extends React.Component {
                 .catch((err) => {
                     console.error(err.stack);
-                    this.setState({ errorMessage: err.response.data })
+                    this.setState({ errorMessage: err.response.data, loginMethodType: this.getVaultAuthMethod() })
```
I think this can be combined
```js
this.setState({
    errorMessage: `${window.suppliedAuthToken ? 
                       'Login was attempted using a server supplied token. Please contact your network administrator -- ' :  ''}${err.response.data}`
});

Commit:
Move logic to constructor and remove unnecessary defaulting

djenriquez commented:
LGTM 👍
